File: test/src/main/java/top/bettercode/summer/test/BaseWebAuthTest.java
package top.bettercode.summer.test;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import top.bettercode.summer.security.ScopeUserDetailsService;
import top.bettercode.summer.security.UserDetailsAuthenticationToken;

/**
 * mockMvc 基础测试类
 *
 * @author Peter Wu
 */
public abstract class BaseWebAuthTest extends BaseWebNoAuthTest {

  protected String username = "root";
  protected String scope = "app";
  @Autowired
  UserDetailsService userDetailsService;

  @Override
  public void defaultBeforeEach() throws Exception {
    beforeEach();
    UserDetails userDetails;
    if (userDetailsService instanceof ScopeUserDetailsService) {
      userDetails = ((ScopeUserDetailsService) userDetailsService).loadUserByScopeAndUsername(scope,
          username);
    } else {
      userDetails = userDetailsService.loadUserByUsername(username);
    }
    SecurityContextHolder.getContext()
        .setAuthentication(new UserDetailsAuthenticationToken(userDetails));
  }


}

File: test/src/main/kotlin/top/bettercode/summer/test/BaseWebAuthTest.kt
package top.bettercode.summer.test

import org.springframework.beans.factory.annotation.Autowired
import top.bettercode.summer.web.support.ApplicationContextHolder

/**
 * mockMvc 基础测试类
 *
 * @author Peter Wu
 */
@Suppress("LeakingThis")
abstract class BaseWebAuthTest : BaseWebNoAuthTest() {

    lateinit var clientId: String
    lateinit var username: String
    lateinit var scope: String

    @Autowired
    lateinit var userDetailsService: TestAuthenticationService

    init {
        this.clientId = ""
        this.username = "root"
        this.scope = "app"
    }

    public override fun defaultBeforeEach() {
        beforeEach()
        if (clientId.isBlank()) {
            this.clientId = ApplicationContextHolder.getProperty("summer.security.client-id", "")
                    ?: ""
        }
        userDetailsService.loadAuthentication(clientId, setOf(scope), username)
    }

}
